pkg/util/config/feature_gate.go

```diff
@@ -31,63 +31,93 @@ const (
 
 	// All known feature keys
 	// To add a new feature, define a key for it below and add
-	// a featureInfo entry to knownFeatures.
-
-	// allAlpha is a global toggle for alpha features. Per feature key
-	// values override the default set by allAlpha.
-	// e.g. allAlpha=false,newFeature=true will result in newFeature=true
-	allAlpha = "allAlpha"
+	// a featureSpec entry to knownFeatures.
+
+	// allAlphaGate is a global toggle for alpha features. Per-feature key
+	// values override the default set by allAlphaGate, if they come later in the
+	// specification of gates. Examples:
+	//   AllAlpha=false,NewFeature=true  will result in newFeature=true
+	//   AllAlpha=true,NewFeature=false  will result in newFeature=false
+	allAlphaGate = "AllAlpha"
 )
 
 var (
-	// DefaultFeatureGate is a shared global FeatureGate.
-	DefaultFeatureGate = &featureGate{}
+	// Default values for recorded features.  Every new feature gate should be
+	// represented here.
+	knownFeatures = map[string]featureSpec{
+		allAlphaGate: {false, alpha},
+	}
 
-	// Default values for recorded features.
-	knownFeatures = map[string]featureInfo{
-		allAlpha: {false, alpha},
+	// Special handling for a few gates.
+	specialFeatures = map[string]func(f *featureGate, val bool){
+		allAlphaGate: setUnsetAlphaGates,
 	}
 
-	alpha = prerelease("ALPHA")
-	beta  = prerelease("BETA")
-	ga    = prerelease("")
+	// DefaultFeatureGate is a shared global FeatureGate.
+	DefaultFeatureGate = &featureGate{
+		known:   knownFeatures,
+		special: specialFeatures,
+	}
 )
 
-type prerelease string
-
-type featureInfo struct {
+type featureSpec struct {
 	enabled    bool
 	prerelease prerelease
 }
 
+type prerelease string
+
+const (
+	// Values for prerelease.
+	alpha = prerelease("ALPHA")
+	beta  = prerelease("BETA")
+	ga    = prerelease("")
+)
+
 // FeatureGate parses and stores flag gates for known features from
 // a string like feature1=true,feature2=false,...
 type FeatureGate interface {
 	AddFlag(fs *pflag.FlagSet)
-	// owner: @jlowdermilk
-	// alpha: v1.4
```

Comment:
I noticed AllAlpha disappeared from the FeatureGate interface.
Do we have a use case for applications calling it directly ? Thinking out loud, I think there isn't a use case, applications should be querying their specific feature gate only.